### PR TITLE
Tie/fix/constrain in multi fitting

### LINKED ISSUE #21829: Tie/fix/constrain in multi fitting
In the multi fitting function browser (muon and general) adding a tie/constraint/fix will be added to all of the workspaces (even if it is not global)

### PR DESCRIPTION
**Pull request by mantid-roman**
Corrected update of the gui for local parameters. Added constraints to the local parameters.

**To test:**

* Start the multi fit interface. Add some data, function.
* Set ties, fixes, constraints to parameters of functions of different data sets.
* Switch between data sets. Check that the gui updates correctly.
* Run a few fits selecting different current data set each time. Results must be the same regardless of the selected data set.

Fixes #21829

**Release Notes** 

*Does not need to be in the release notes.*

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [x] Is the code of an acceptable quality?
- [x] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [x] Do changes function as described? Add comments below that describe the tests performed?
- [x] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by keeeto** [comment]
I'm happy to have a go at testing this too. Can anyone direct me to the multi fitting documentation?

**Comment by mantid-roman** [comment]
@keeeto I am not sure if we have it

**Comment by keeeto** [comment]
Ah, okay. No problem; I guess this is a job I cannot do then :). Maybe after the release is over I will have a go at learning the interface and writing some documentation. @mantid-roman would I be best to talk to you about this?

**Comment by mantid-roman** [comment]
@keeeto Yes

**Comment by soininen** [comment]
I found a problem when using CompositeFunction.
1. Get some data. I used `084446.nxs` from `ExternalData/Testing/Data/UnitTest/ILL/IN4`.
2. Open the Multi dataset fitting interface.
3. Add data. I used ws indices 10-14 from the IN4 workspace.
4. Add composite function.
5. Add FlatBackground to the composite function.
6. Constrain the A0 parameter to 10%.
7. Click Fit. Mantid crashes.

There is even a crash if I just try to change the value of A0...

**Comment by soininen** [comment]
The problem with CompositeFunction seems to be gone now. However, there are some problems with ties: when I write some garbage, like `0.8*` in the tie input field, Mantid crashes.

Also, the tie field sometimes becomes disabled (grayed out) and I cannot modify the value anymore. This seems to happen when I try to write `0.` in the field.

**Comment by soininen** [comment]
I should have been more precise: there is a crash when I right click a parameter, select `Add tie`, and write some garbage like `**` in the `Add a tie` dialog.

**Comment by soininen** [comment]
The problems with the ties are now gone.

However, if I
1. Use the same dataset as above
2. Create a CompositeFunction
3. Add a FlatBackground
4. Add a Gaussian
5. Set the tie of the Sigma parameter to `f1.PeakCentre/25`
6. Click 'Fit'

```
Fit failed:

Unknown function parameter Height

This may happen if there is a CompositeFunction containing only one function
```

**Comment by mantid-roman** [comment]
I am afraid this is unfixable, in this release at least. The message tells the user that using a composite function with one member is an error and they should either add more functions of do without the CompositeFunction.

**Comment by soininen** [comment]
@mantid-roman There was two functions there in this case, FlatBackground and Gaussian, so I was a bit surprised with the error message.

**Comment by mantid-roman** [comment]
That's not good :(

File: qt/widgets/common/src/FunctionBrowser.cpp [hunk 1]
```diff
@@ -1912,47 +1953,72 @@ void FunctionBrowser::attributeVectorDoubleChanged(QtProperty *prop) {
...
-      return;
-    auto tie = QString::fromStdString(getTie(parProp));
-    auto tieExpr = tie.split('=');
-    if (tieExpr.size() == 2) {
-      tie = tieExpr[1];
+/// Called when a constraint property changes
+void FunctionBrowser::constraintChanged(QtProperty *prop) {
+  for (auto it = m_constraints.begin(); it != m_constraints.end(); ++it) {
+    bool isLower = it.value().lower == prop;
+    bool isUpper = it.value().upper == prop;
```
Comment (soininen) [inline qt/widgets/common/src/FunctionBrowser.cpp]: `const bool`?

File: qt/widgets/common/src/FunctionBrowser.cpp [hunk 2]
```diff
@@ -1912,47 +1953,72 @@ void FunctionBrowser::attributeVectorDoubleChanged(QtProperty *prop) {
...
+    if (isLower || isUpper) {
+      auto paramProp = it.value().paramProp;
+      if (isLocalParameterProperty(paramProp)) {
+        auto parName = getParameterName(paramProp);
```
Comment (soininen) [inline qt/widgets/common/src/FunctionBrowser.cpp]: `const auto`?

File: qt/widgets/common/src/FunctionBrowser.cpp [hunk 3]
```diff
@@ -2171,7 +2250,8 @@ Mantid::API::IFunction_sptr FunctionBrowser::getGlobalFunction() {
     multiFun->setDomainIndex(i, i);
     auto fun1 = multiFun->getFunction(i);
     for (size_t j = 0; j < fun1->nParams(); ++j) {
-      QString parName = QString::fromStdString(fun1->parameterName(j));
+      auto parameterName = fun1->parameterName(j);
+      QString parName = QString::fromStdString(parameterName);
```
Comment (soininen) [inline qt/widgets/common/src/FunctionBrowser.cpp]: Could both strings be `const`? It looks a bit strange that the same string is duplicated as C-string and `QString` here, but I guess it cannot be helped.

File: qt/widgets/common/src/FunctionBrowser.cpp [hunk 4]
```diff
@@ -2181,19 +2261,32 @@ Mantid::API::IFunction_sptr FunctionBrowser::getGlobalFunction() {
...
+      if (localParam.fixed) {
         // Fix this particular local parameter
-        fun1->tie(parName.toStdString(),
-                  boost::lexical_cast<std::string>(
-                      getLocalParameterValue(parName, i)));
+        fun1->setParameter(j, localParam.value);
+        fun1->fix(j);
       } else {
-        auto tie = getLocalParameterTie(parName, i);
+        auto tie = localParam.tie;
```
Comment (soininen) [inline qt/widgets/common/src/FunctionBrowser.cpp]: `const auto &`?

File: qt/widgets/common/src/FunctionBrowser.cpp [hunk 5]
```diff
@@ -1912,47 +1953,72 @@ void FunctionBrowser::attributeVectorDoubleChanged(QtProperty *prop) {
...
-    auto &paramValue = m_localParameterValues[parName][m_currentDataset];
-    if (paramValue.fixed)
-      return;
-    auto tie = QString::fromStdString(getTie(parProp));
-    auto tieExpr = tie.split('=');
-    if (tieExpr.size() == 2) {
-      tie = tieExpr[1];
+/// Called when a constraint property changes
+void FunctionBrowser::constraintChanged(QtProperty *prop) {
+  for (auto it = m_constraints.begin(); it != m_constraints.end(); ++it) {
```
Comment (soininen) [inline qt/widgets/common/src/FunctionBrowser.cpp]: Perhaps C++11 style `for` loop could be considered here:
```
for (const auto &constraint : m_constraints) {
```

**Review by soininen — CHANGES_REQUESTED** [review]
Tested the changes with some data, and for simple functions, ties, fixes and constraints work fine. I get some error message with ties when CompositeFunction is involved but this is a known problem to my understanding.

The code changes look good, as well. I spotted just a few minor issues.

**Comment by soininen** [comment]
The ties seem to work now with multiple functions in CompositeFunction. Good work!

**Review by soininen — APPROVED** [review]

**Pull request merged**